scripts/notebooks/Batching.ipynb

```diff
@@ -0,0 +1,1275 @@
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "INCLUDE_BINS = ['TODO']\n",
+    "\n",
+    "validate_include_bins(INCLUDE_METRICS, INCLUDE_BINS)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-block alert-success\">Input the target batch size you'd like to create, as well as the minimum and maximum batch sizes.</div>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "TARGET_BATCH_SIZE = 'TODO'\n",
+    "MINIMUM_BATCH_SIZE = 'TODO'\n",
+    "MAXIMUM_BATCH_SIZE = 'TODO'\n",
+    "\n",
+    "validate_numeric_inputs([TARGET_BATCH_SIZE, MINIMUM_BATCH_SIZE, MAXIMUM_BATCH_SIZE])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-block alert-success\">Input the batch prefixes and suffixes you'd like to use. Remember to adhere to the <a href=\"https://broadinstitute.github.io/gatk-sv/docs/gs/inputs#sampleids\">naming requirements</a>.</div>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "BATCH_PREFIX = \"batch\"\n",
+    "BATCH_SUFFIX = \"\"\n",
+    "\n",
+    "validate_string_inputs([BATCH_PREFIX, BATCH_SUFFIX])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-block alert-success\">If you wish to group related samples in the same batch, set PED_FILE to True - if not, then leave this as is. Note that batching related samples together may result in less homogeneous batches.</div>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "PED_FILE = False\n",
+    "\n",
+    "reference_ped = None\n",
+    "default_ped_path = os.path.join(WS_BUCKET, \"evidence_qc/sex_analysis/sample_qc.ped\")\n",
+    "if PED_FILE:\n",
+    "    validate_ped(default_ped_path, set(pass_df['sample_id']))\n",
+    "    reference_ped = pd.read_table(default_ped_path, dtype=str, header=None, names = [\n",
+    "        \"Family_ID\", \"Sample_ID\", \"Paternal_ID\", \"Maternal_ID\", \"Sex\", \"Phenotype\"\n",
+    "    ])\n",
+    "    reference_ped = reference_ped[reference_ped['Sample_ID'].isin(pass_df['sample_id'])]\n",
+    "else:  \n",
+    "    print(\"PED file is not provided - skipping this step.\")\n",
+    "\n",
+    "reference_ped"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create Batches"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Perform hierarchical batching and print summary of batch size and sex balance\n",
+    "batches, batches_meta = generate_hierarchical_batches(pass_df, INCLUDE_METRICS, INCLUDE_BINS, TARGET_BATCH_SIZE, \n",
```

Comment:
This is getting stuck in an infinite loop with target size 500, max 600, min 100

Reply:
Do you mind sharing the workspace you encountered this in with me? It would be helpful to trigger the error myself, as I'm not 100% sure about the underlying source of this.

Reply:
I just uploaded a version with more explicit checks on batch sizes - I believe the issue was being triggered because you expected 10 splits, yet only had ~150 samples. This does not allow for valid batches to be created based on the batch size thresholds provided.